mesh/ncnurbs.cpp

```diff
@@ -0,0 +1,3344 @@
+
+   const int i0 = f00[0];
+   const int j0 = f00[1];
+
+   for (int i=0; i<n1; ++i)
+      for (int j=0; j<n2; ++j)
+      {
+         // Entity perm[i] of the face should be entity i in the master face
+         // ordering. The master face ordering varies faster in the direction
+         // from v0 to v10, and slower in the direction from v10 to v11, or
+         // equivalently, from v0 to v01.
+         if (tr)
+         {
+            const int fi = i0 == 0 ? j : n2 - 1 - j;
+            const int fj = j0 == 0 ? i : n1 - 1 - i;
+            const int p = fi + (fj * n2); // Index in the slave face ordering
+            const int m = i + (j * n1); // Index in the master face ordering
+            perm[m] = p;
+         }
+         else
+         {
+            const int fi = i0 == 0 ? i : n1 - 1 - i;
+            const int fj = j0 == 0 ? j : n2 - 1 - j;
+            const int p = fi + (fj * n1); // Index in the slave face ordering
+            const int m = i + (j * n1); // Index in the master face ordering
+            perm[m] = p;
+         }
+      }
+}
+
+bool ConsistentlySetEntry(int v, int &e)
+{
+   const bool consistent = e == -1 || e == v;
+   e = v;
+   return consistent;
+}
+
+void ReorderArray2D(int i0, int j0, const Array2D<int> &a,
+                    Array2D<int> &b)
+{
+   const int m = a.NumRows();
+   const int n = a.NumCols();
+
+   b.SetSize(m, n);
+
+   const int s0 = i0 == 0 ? 1 : -1;
+   const int s1 = j0 == 0 ? 1 : -1;
+   for (int i=0; i<m; ++i)
+   {
+      const int ia = (i0 * (m - 1)) + (s0 * i);
+      for (int j=0; j<n; ++j)
+      {
+         const int ja = (j0 * (n - 1)) + (s1 * j);
+         b(i, j) = a(ia, ja);
+      }
+   }
+}
+
+void GetVertexOrdering(int ori, std::array<int, 4> &perm)
+{
+   const int oriAbs = ori < 0 ? -1 - ori : ori;
+
+   for (int i=0; i<4; ++i)
+   {
+      if (ori < 0)
+      {
+         perm[i] = (oriAbs - i + 4) % 4;
+      }
+      else
+      {
+         perm[i] = (ori + i) % 4;
+      }
+   }
+}
+
+// Append master face DOFs to masterDofs.
+void NURBSPatchMap::SetMasterFaces(bool dof)
+{
+   faceMaster.SetSize(faces.Size());
+   faceMasterOffset.SetSize(faces.Size());
+
+   // The loop over master edges is already done by SetMasterEdges, and now we
+   // append face DOFs to masterDofs.
+
+   int mos = masterDofs.Size();
+   for (int i=0; i<faces.Size(); ++i)
+   {
+      faceMaster[i] = Ext->masterFaces.count(faces[i]) > 0;
+      faceMasterOffset[i] = mos;
+
+      if (!faceMaster[i]) { continue; }
+
+      Array2D<int> mdof;
+      Ext->GetMasterFaceDofs(dof, faces[i], mdof);
+      if (mdof.NumRows() == 0)
+      {
+         faceMaster[i] = false;
+         continue;
+      }
+
+      for (int j=0; j<mdof.NumCols(); ++j)
+         for (int k=0; k<mdof.NumRows(); ++k)
+         {
+            masterDofs.Append(mdof(k,j));
+         }
+
+      mos += mdof.NumRows() * mdof.NumCols();
+   } // loop (i) over faces
+}
+
+void NURBSExtension::GetMasterFaceDofs(bool dof, int mf,
```

Comment:
Also a very long function. Perhaps split it into smaller ones with their own functionality.

Reply:
Note that a significant portion of this function is the lambda function `SetEdgeEntries`. I don't see a way to decompose further.